src/modules/schnorrsig/main_impl.h

```diff
@@ -264,4 +264,206 @@ int secp256k1_schnorrsig_verify(const secp256k1_context* ctx, const unsigned cha
+        secp256k1_sha256_finalize(&hashcopy, hashoutput);
+        secp256k1_scalar_set_b32(&zi, hashoutput, NULL);
+
+        /* Step2: s := s + zi*si */
+        secp256k1_scalar_set_b32(&si, &new_sigs64[(i-n_before)*64+32], NULL);
+        secp256k1_scalar_mul(&si, &si, &zi);
+        secp256k1_scalar_add(&s, &s, &si);
+    }
+
+    /* copy new rs into aggsig */
+    for (i = n_before; i < n; ++i) {
+        memcpy(&aggsig[i*32], &new_sigs64[(i-n_before)*64], 32);
+    }
+    /* copy new s into aggsig */
+    secp256k1_scalar_get_b32(&aggsig[n*32], &s);
+    *aggsig_size = 32 * (1 + n);
+    return 1;
+}
+
+int secp256k1_schnorrsig_aggregate(const secp256k1_context* ctx, unsigned char* aggsig, size_t* aggsig_size, const secp256k1_xonly_pubkey* pubkeys, const unsigned char* msgs32, const unsigned char* sigs64, size_t n) {
+    if (!secp256k1_schnorrsig_inc_aggregate(ctx, aggsig, aggsig_size, pubkeys, msgs32, sigs64, 0, n)) {
+        return 0;
+    }
+    return 1;
+}
+
+int secp256k1_schnorrsig_aggverify(const secp256k1_context* ctx, const secp256k1_xonly_pubkey* pubkeys, const unsigned char* msgs32, size_t n, const unsigned char* aggsig, size_t aggsig_size) {
+    size_t i;
+    secp256k1_gej lhs, rhs;
+    secp256k1_scalar s;
+    secp256k1_sha256 hash;
+    int overflow;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(pubkeys != NULL);
+    ARG_CHECK(msgs32 != NULL);
+    ARG_CHECK(aggsig != NULL);
+
+    /* Check that aggsig_size is correct, i.e. aggsig_size = 32*(n+1) */
+    if ((aggsig_size >> 5) <= 0 || ((aggsig_size >> 5)-1) != n) {
+        return 0;
+    }
+
+    /* Compute the rhs:               */
+    /* Set rhs = 0                    */
+    /* For each i in 0,.., n-1, do:   */
+    /*     (1) z_i = TaggedHash(...)  */
+    /*     (2) T_i = R_i+e_i*P_i      */
+    /*     (3) rhs = rhs + z_i*T_i    */
+    secp256k1_gej_set_infinity(&rhs);
+    secp256k1_schnorrsig_sha256_tagged_aggregation(&hash);
+    for (i = 0; i < n; ++i) {
+        secp256k1_fe rx;
+        secp256k1_ge rp, pp;
+        secp256k1_scalar ei;
+        secp256k1_gej ppj, acc;
+
+        unsigned char pk_ser[32];
+        unsigned char hashoutput[32];
+        secp256k1_sha256 hashcopy;
+        secp256k1_scalar zi;
+
+        /* Step 0: Serialize pk_i into pk_ser   */
+        /* We need that in Step 1 and in Step 2 */
+        if (!secp256k1_xonly_pubkey_serialize(ctx, pk_ser, &pubkeys[i])) {
+            return 0;
+        }
+
+        /* Step 1: z_i = TaggedHash(...) */
+        /* 1.a) Write into hash r_i, pk_i, mes_i, r_i */
+        secp256k1_sha256_write(&hash, &aggsig[i*32], 32);
+        secp256k1_sha256_write(&hash, pk_ser, 32);
+        secp256k1_sha256_write(&hash, &msgs32[i*32], 32);
+        /* 1.b) Copy the hash */
+        hashcopy = hash;
+        /* 1.c) Finalize the copy to get zi*/
+        secp256k1_sha256_finalize(&hashcopy, hashoutput);
+        secp256k1_scalar_set_b32(&zi, hashoutput, NULL);
```

Comment:
Have you considered extracting this to a separate helper function? It's also used in signing. (I am not entirely sure if it will be more readable in the end.)

Reply:
I think it will be less readable in the end. The function would not only have to output the zi, but also update hash.